docs/stability-wasi-proposals-support.md

```diff
@@ -0,0 +1,38 @@
+# WASI Proposals Support
+
+Wasmtime has initial support for some WASI proposals to assist in their
+development as well as allowing users to more easily provide feedback by testing
+them. This is somewhat of a work-in-progress because there aren't a lot of WASI
+proposals at this time.
+
+Currently supported WASI proposals are:
+
+* [The base WASI specification](https://github.com/webassembly/wasi) - the
+  `snapshot` directory is supported by the [`wasmtime-wasi`
+  crate](https://docs.rs/wasmtime-wasi/0.21.0/wasmtime_wasi/)
+
+* [The wasi-nn proposal](https://github.com/WebAssembly/wasi-nn) - the
+  `ephemeral` directory with the wasi-nn APIs.
+
+Wasmtime strives to support WASI proposals in-tree to assist with
+maintainability and provide easier to support to users for easier feedback. In
+doing so, however, we have a few guidelines around adding support for a WASI
+proposal to Wasmtime:
+
+* The proposal must be tested and exercised on CI. These tests don't need to be
+  overly exhaustive but a "smoke test" should pass. Additionally CI should not
+  be overly long compared to existing tests run, and it should also strive to be
+  relataively reliable. Wasmtime contributors may disable CI for proposals at
+  any time if needed, and maintainers for the proposal will be notified.
+
+* The size of the code in the wasmtime repository should not be unduly large.
+  This helps keep the size of the repo within reasonable limits and relatively
+  easy for others not working on the proposal to still be productive.
+
+* It should be possible to compile wasmtime without support for the proposal.
+  It's ok to compile it in by default (but this may not always be the right
+  choice), but Wasmtime should always be able to compile without the proposal.
+  Additionally the proposal doesn't need to support all platforms in the get-go.
```

Comment:
We should also have guidelines for whether features should be enabled by default at runtime, or require a flag to be set.

Maybe we can differentiate some of this by which stage in the process the proposal is in? Something like
- stage 0 proposals must be disabled by default at compile time, and must require a flag at runtime
- stage 1 proposals can be enabled by default, but must require a flag at runtime
- stage 2+ proposals can be enabled by default at compile and runtime

Reply:
An additional guideline might be about which namespace the API should be in. Perhaps stage 2+ can use the WASI namespacing system, but everything before some explicitly more experimental namespace? @sunfishcode and @pchickey, I'd be curious what your take is on this?

Reply:
Lets leave the namespace up to wherever the proposal is with in the WASI subgroup, but also permit an experimental namespace to be used by all proposals.

Reply:
Good point that this is a WASI subgroup topic more than anything else. So perhaps a more general statement could be included here around following the WASI subgroup's rules around exposure of in-progress proposals?